src/transports/http.rs

```diff
@@ -86,16 +91,15 @@ async fn execute_rpc<T: DeserializeOwned>(client: &Client, url: Url, request: &R
         .bytes()
         .await
         .map_err(|err| Error::Transport(format!("failed to read response bytes: {}", err)))?;
-    log::debug!(
-        "[id:{}] received response: {:?}",
-        id,
-        String::from_utf8_lossy(&response).as_ref()
-    );
+    let decoded_response = String::from_utf8_lossy(&response);
+    log::debug!("[id:{}] received response: {:?}", id, decoded_response.as_ref());
     if !status.is_success() {
-        return Err(Error::Transport(format!(
-            "response status code is not success: {}",
-            status
-        )));
+        let code = ErrorCode::from(i64::from(status.as_u16()));
+        return Err(Error::Rpc(RPCError {
+            code,
```

Comment:
Not a fan of mixing HTTP error codes (i.e. transport specific) with the ones returned by actual JSON-RPC server. What if we have a conflict and conflate an JSON-RPC error code with some HTTP error code?

I'd rather introduce a richer `Transport` variant, with following fields:
```rust
struct TransportError {
  /// Transport implementation self-identification.
  /// 
  /// This can be used to determine the actual transport (HTTP, WS, etc). being used if it's not statically known.
  transport_id: String,
  /// Transport-specific error code.
  code: u64,
  /// Arbitrary, developer-readable description of the occurred error.
  message: String,
}
```
wdyt?

Reply:
JSON-RPC codes are to be negative, http/ws error codes are positive so there shouldn't be a clash, so I'm fine with both approaches. It makes some sense to use a different variant for transport and rpc errors because the former are much more retriable than the latter in many cases.

This was roughly my initial idea, however, `Transport` is not used in a compatible way right now as it's also used to report errors like https://github.com/tomusdrw/rust-web3/blob/033ac37fff3990a943b9362608cd2c151914f3e9/src/transports/http.rs#L51 or https://github.com/tomusdrw/rust-web3/blob/033ac37fff3990a943b9362608cd2c151914f3e9/src/transports/http.rs#L88. 

We could either introduce a new variant or use `code: Option<u64>`.

Reply:
> JSON-RPC codes are to be negative, http/ws error codes are positive so there shouldn't be a clash

JSON-RPC codes can be anything actually, the spec just defines a few reserved ranges which are negative.

> This was roughly my initial idea, however, Transport is not used in a compatible way right now as it's also used to report errors like

We can either assign a `code` to each of such error (and expose constants with that values), use a "catch-all" `0` code for that or use `Option<u64>` as you suggested - I'm fine with either.

Reply:
I did a slightly different implementation given that we're unlikely to get a developer-defined message and a code. I also skipped the `transport_id` because it would be a lot of boilerplate.